shadowsocks/tcp.go

```diff
@@ -199,29 +210,28 @@ func (s *tcpService) Start() {
 			clientConn = metrics.MeasureConn(clientConn, &proxyMetrics.ProxyClient, &proxyMetrics.ClientProxy)
 			defer func() {
 				connDuration := time.Now().Sub(connStart)
-				clientConn.Close()
 				status := "OK"
 				if connError != nil {
 					logger.Debugf("TCP Error: %v: %v", connError.Message, connError.Cause)
 					status = connError.Status
 				}
-				logger.Debugf("Done with status %v, duration %v", status, connDuration)
 				s.m.AddClosedTCPConnection(clientLocation, keyID, status, proxyMetrics, timeToCipher, connDuration)
+				clientConn.Close() // Closing after the metrics are added aids integration testing.
+				logger.Debugf("Done with status %v, duration %v", status, connDuration)
 			}()
 
 			findStartTime := time.Now()
-			keyID, clientConn, err := findAccessKey(clientConn, *s.ciphers)
+			keyID, clientConn, salt, err := findAccessKey(clientConn, *s.ciphers)
 			timeToCipher = time.Now().Sub(findStartTime)
 
 			if err != nil {
-				// Keep the connection open until we hit the authentication deadline to protect against probing attacks
 				logger.Debugf("Failed to find a valid cipher after reading %v bytes: %v", proxyMetrics.ClientProxy, err)
-				_, drainErr := io.Copy(ioutil.Discard, clientConn) // drain socket
-				drainResult := drainErrToString(drainErr)
-				port := s.listener.Addr().(*net.TCPAddr).Port
-				logger.Debugf("Drain error: %v, drain result: %v", drainErr, drainResult)
-				s.m.AddTCPProbe(clientLocation, drainResult, port, proxyMetrics)
+				s.absorbProbe(clientConn, clientLocation, &proxyMetrics)
 				return onet.NewConnectionError("ERR_CIPHER", "Failed to find a valid cipher", err)
+			} else if !s.replayCache.Add(salt) { // Only check the cache if findAccessKey succeeded.
+				logger.Debug("Replay detected")
```

Comment:
Let's print the salt. I think that will be informative.

Reply:
Done